src/core/cmcd/cmcd_data_builder.ts

```diff
@@ -0,0 +1,602 @@
+  IReadOnlyPlaybackObserver,
+  IRebufferingStatus,
+  ObservationPosition,
+} from "../../playback_observer";
+import type { ICmcdOptions, ICmcdPayload, ITrackType } from "../../public_types";
+import createUuid from "../../utils/create_uuid";
+import isNullOrUndefined from "../../utils/is_null_or_undefined";
+import type { IRange } from "../../utils/ranges";
+import TaskCanceller from "../../utils/task_canceller";
+
+/**
+ * `rtp`, for "REQUESTED_MAXIMUM_THROUGHPUT", indicates the maximum throughput
+ * needed to load a given segment without experience degration.
+ * It acts as a hint to a CDN so it can scale its resources between multiple
+ * clients.
+ *
+ * We could indicate through `rtp` the exact minimum bandwidth needed, but this
+ * may lead to much higher risk of rebuffering, so we prefer to multiply that
+ * value by a safe-enough factor, this `RTP_FACTOR`.
+ */
+const RTP_FACTOR = 4;
+
+/**
+ * Information that should be provided to the `CmcdDataBuilder` when getting the
+ * CMCD payload for a segment.
+ */
+export interface ICmcdSegmentInfo {
+  /** Manifest metadata linked to the wanted segment. */
+  manifest: IManifest;
+  /** Period metadata linked to the wanted segment. */
+  period: IPeriod;
+  /** Adaptation metadata linked to the wanted segment. */
+  adaptation: IAdaptation;
+  /** Representation metadata linked to the wanted segment. */
+  representation: IRepresentation;
+  /** Segment metadata linked to the wanted segment. */
+  segment: ISegment;
+}
+
+/**
+ * Media playback observation's properties the `CmcdDataBuilder` wants to have
+ * access to.
+ */
+export interface ICmcdDataBuilderPlaybackObservation {
+  /**
+   * Ranges of buffered data per type of media.
+   * `null` if no buffer exists for that type of media.
+   */
+  buffered: Record<ITrackType, IRange[] | null>;
+  /**
+   * Information on the current media position in seconds at the time of the
+   * Observation.
+   */
+  position: ObservationPosition;
+  /** Target playback rate at which we want to play the content. */
+  speed: number;
+  /**
+   * Describes when the player is "rebuffering" and what event started that
+   * status.
+   * "Rebuffering" is a status where the player has not enough buffer ahead to
+   * play reliably.
+   * The RxPlayer should pause playback when a playback observation indicates the
+   * rebuffering status.
+   */
+  rebuffering: IRebufferingStatus | null;
+}
+
+/**
+ * Class allowing to easily obtain "Common Media Client Data" (CMCD) properties
+ * that may be relied on while performing HTTP(S) requests on a CDN.
+ *
+ * @class CmcdDataBuilder
+ */
+export default class CmcdDataBuilder {
+  private _sessionId: string;
+  private _contentId: string;
+  private _typePreference: TypePreference;
+  private _lastThroughput: Partial<Record<ITrackType, number | undefined>>;
+  private _playbackObserver: IReadOnlyPlaybackObserver<ICmcdDataBuilderPlaybackObservation> | null;
+  private _bufferStarvationToggle: boolean;
```

Comment:
It switches on when buffering - until the `bs` (buffer starvation) attribute is constructed, at which point it is reset to `false`.

That's the only attribute which necessitates some track-keeping of what has been sent previously - which is kind of problematic to us by the way because knowing whether we actually reached the CDN previously (and not some cache, some custom segment loaders, local file creation etc.) is hard.

Reply:
Ok it make sense with  your explanation